### PR TITLE
Deferred for custom slot

### PR DESCRIPTION
**Pull request by antmat**
Deffered is quite usefull even for custom protocol transitions. This patch allows defferred and streamed to be used in custom slots.
Also some const callables such as std::mem_fn can call only underlying const methods. So in general we might want to pass reference instead of const reference to tuple::invoke. This patch only allows passing rvalue references additionally to lvalue const references to be compatible with old behaviour.

### CODE REVIEW

File: include/cocaine/rpc/queue.hpp [hunk 1]
```diff
@@ -90,8 +90,10 @@ struct frozen_visitor:
 
 template<class Tag, class Upstream>
 class message_queue {
+public:
     typedef Upstream upstream_type;
```
Comment (Kobolog) [inline include/cocaine/rpc/queue.hpp]: This is not used anywhere in this PR, why did you change it to public?

File: include/cocaine/rpc/slot/deferred.hpp [hunk 2]
```diff
@@ -128,6 +128,12 @@ struct deferred {
         return *this;
     }
 
+    template <class upstream_type>
```
Comment (Kobolog) [inline include/cocaine/rpc/slot/deferred.hpp]: Codestyle: template parameters are single capital letters (e.g. `U`) or CamelCase (e.g. `UpstreamType`). No space between `template` and an opening `<`.

File: include/cocaine/rpc/slot/deferred.hpp [hunk 3]
```diff
@@ -157,6 +163,11 @@ struct deferred<void> {
...
```
Comment (Kobolog) [inline include/cocaine/rpc/slot/deferred.hpp]: Same here.

File: include/cocaine/rpc/slot/streamed.hpp [hunk 4]
```diff
@@ -60,6 +60,11 @@ struct streamed {
         return *this;
     }
 
+    template <class upstream_type>
```
Comment (Kobolog) [inline include/cocaine/rpc/slot/streamed.hpp]: Same here.

**Comment by Kobolog** [comment]
Also, please elaborate a bit on rvalue-reference callables in tuple.hpp, preferably with some examples.

File: include/cocaine/rpc/queue.hpp (hunk 1 above)
Comment (antmat) [inline include/cocaine/rpc/queue.hpp]: My fault. I was thinking about making non template attach with type defined in message_queue but changed my mind and did not change public back to private.

File: include/cocaine/rpc/queue.hpp [hunk 5]
```diff
@@ -92,6 +92,7 @@ template<class Tag, class Upstream>
 class message_queue {
     typedef Upstream upstream_type;
 
+private:
```
Comment (Kobolog) [inline include/cocaine/rpc/queue.hpp]: This is redundant too, private by default.

File: include/cocaine/rpc/slot/deferred.hpp [hunk 6]
```diff
@@ -128,6 +128,12 @@ struct deferred {
         return *this;
     }
 
+    template<class UpstreamType>
+    void attach(UpstreamType&& upstream) const {
```
Comment (Kobolog) [inline include/cocaine/rpc/slot/deferred.hpp]: Codestyle (missed it the first time, sorry): return types should be on a separate line, i.e.:

```
template<class UpstreamType>
void
attach(UpstreamType&& upstream) const { ...
```

File: include/cocaine/rpc/slot/deferred.hpp [hunk 7]
```diff
@@ -157,6 +163,11 @@ struct deferred<void> {
         return *this;
     }
 
+    template<class UpstreamType>
+    void attach(UpstreamType&& upstream) {
```
Comment (Kobolog) [inline include/cocaine/rpc/slot/deferred.hpp]: Here too.

File: include/cocaine/rpc/slot/streamed.hpp [hunk 8]
```diff
@@ -60,6 +60,11 @@ struct streamed {
         return *this;
     }
 
+    template<class UpstreamType>
+    void attach(UpstreamType&& upstream) const {
```
Comment (Kobolog) [inline include/cocaine/rpc/slot/streamed.hpp]: Here too.

**Comment by antmat** [comment]
I've made changes as requested. Example of rvalue-reference callables can be found in cocaine/cocaine-plugins#52 (and as mentioned - it's blocked by this PR)
